certora/specs/Delegation.spec

```
@@ -0,0 +1,61 @@
+import "ERC20.spec";
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sload uint256 votingPower (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] {
+    require sumOfVotingPower >= to_mathint(votingPower);
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] uint256 newValue (uint256 oldValue) {
+    sumOfVotingPower = sumOfVotingPower - oldValue + newValue;
+}
+
+// Check that zero address has no voting power assuming that zero address can't make transactions.
+invariant zeroAddressNoVotingPower()
+    delegatee(0x0) == 0x0 && delegatedVotingPower(0x0) == 0
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    { preserved with (env e) { require e.msg.sender != 0; } }
+
+// Check that the voting power is never greater than the total supply of tokens.
+invariant totalSupplyLTEqSumOfVotingPower()
+    to_mathint(totalSupply()) == sumOfVotingPower + currentContract._zeroVirtualVotingPower
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    {
+      preserved  with (env e) {
+          requireInvariant totalSupplyIsSumOfBalances();
+          requireInvariant zeroAddressNoVotingPower();
+      }
+    }
+
+// Check that user can restore their voting power by delegating to zero address then delgating back to themselves.
```

```suggestion
// Check that users can benefit from their voting power by delegating to themselves.
```
Could we try to verify this instead ? It's a bit more general: the user delegates to someone else, instead of starting with a delegation to 0

---

Initially, I chose the word "restore" as it conveyed that no matter what you do with your tokens, you can always get your voting power back by delegating to zero then to yourself. Also, this rule not only proves that but also that no new voting power has been created out of the blue. This why the rule name has the suffix consistent, i.e. consistent with respect to the voting power, although we could argue that this also follows from the invariant on the sum of voting power. 

I am under the impression that the rule that you're suggesting is different. Also, I think your current statement misses three facts:
 - the delegatee of the user may have delegated his votes, so we can't simply states the property with respect to the delegatee voting power; 
 - targeting the delegatee of the delegatee is not enough as he may have delegated to zero;
 - the initially the fact `delegatee(user) == delgatee(delegatee)` may hold and therefore the voting power remains unchanged.

 These fact makes you original statement false, doesn't it?

---

> you can always get your voting power back by delegating to zero then to yourself

why do you need to delegate to zero first ? Can't you just delegate to yourself directly ?

> the delegatee may be zero, in which case the balance is unchanged;

This is the same as in the current rule right ? The initial delegatee is zero, and this rule seems to go through

> initially the fact delegator(user) == delegatee may hold and therefore the voting power remains unchanged.

This would be handled the same as in the current rule. So instead of 
```solidity
require user != 0;
```
we do
```solidity
require user != previousDelegatee;
```

---

Fixed in https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/f80c7fbaebcbaf0f9b92f71aeafeed59a04c15fc.